syntax = "proto3";
package blog;



// services
service BlogService {
    rpc CreateBlog(CreateBlogReq) returns (CreateBlogRes);
    rpc ReadBlog(ReadBlogReq) returns (ReadBlogRes);
    rpc UpdateBlog(UpdateBlogReq) returns (UpdateBlogRes);
    rpc DeleteBlog(DeleteBlogReq) returns (DeleteBlogRes);
    rpc ListBlogs(ListBlogRequest) returns (stream ListBlogResponse);
    rpc SearchBlog(SearchRequest) returns (stream SearchResponse);
    rpc GetByID(GetByIDRequest) returns (stream GetByIDResponse);
    rpc GetByAuthorID(GetByAuthorIDRequest) returns (GetByAuthorIDResponse);
}


// messages
message Blog {
    string id = 1;
    string author_id = 2;
    string title = 3;
    string content= 4;
}

message CreateBlogReq {
    Blog blog = 1; // Blog id blank
}

message CreateBlogRes {
    Blog blog = 1; // Blog id filled in
}

message UpdateBlogReq {
    Blog blog = 1;
}

message UpdateBlogRes {
    bool success = 1;
    string msg = 2;
}

message ReadBlogReq {
    string id = 1;
}

message ReadBlogRes {
    Blog blog = 1;
}

message DeleteBlogReq {
    string id = 1;
}

message DeleteBlogRes {
    bool success = 1;
    string msg = 2;
}

message ListBlogRequest {}

message ListBlogResponse {
    Blog blog = 1;
}

message SearchRequest {
    string query = 1;
}

message SearchResponse {
    Blog blog = 1;
    bool status = 2;
}

message GetByIDRequest {
    string id = 1;
}

message GetByIDResponse {
    Blog blog = 1;
}

message GetByAuthorIDRequest {
    string author_id = 1;
}

message GetByAuthorIDResponse {
    Blog blog = 1;
}